smac/epm/rf_with_instances.py

```diff
@@ -195,21 +195,36 @@ def _predict(self, X: np.ndarray) -> typing.Tuple[np.ndarray, np.ndarray]:
         if X.shape[1] != self.types.shape[0]:
             raise ValueError('Rows in X should have %d entries but have %d!' % (self.types.shape[0], X.shape[1]))
 
-        means, vars_ = [], []
-        for row_X in X:
-            if self.log_y:
+        if self.log_y:
+            all_preds = []
+            third_dimension = 0
+
+            # Gather data in a list of 2d arrays and get statistics about the required size of the 3d array
+            for row_X in X:
                 preds_per_tree = self.rf.all_leaf_values(row_X)
-                means_per_tree = []
-                for preds in preds_per_tree:
-                    # within one tree, we want to use the
-                    # arithmetic mean and not the geometric mean
-                    means_per_tree.append(np.log(np.mean(np.exp(preds))))
-                mean = np.mean(means_per_tree)
-                var = np.var(means_per_tree) # variance over trees as uncertainty estimate
-            else:
-                mean, var = self.rf.predict_mean_var(row_X)
-            means.append(mean)
-            vars_.append(var)
+                all_preds.append(preds_per_tree)
+                max_num_leaf_data = max(map(len, preds_per_tree))
+                third_dimension = max(max_num_leaf_data, third_dimension)
+
+            # Transform list of 2d arrays into a 3d array
+            preds_as_array = np.empty((X.shape[0], self.rf_opts.num_trees, third_dimension)) * np.NaN
+            for i, preds_per_tree in enumerate(all_preds):
+                for j, pred in enumerate(preds_per_tree):
+                    preds_as_array[i, j, :len(pred)] = pred
+
+            # Do all necessary computation with vectorized functions
+            preds_as_array = np.log(np.nanmean(np.exp(preds_as_array), axis=2) + 1e-10)
```

Comment:
why have you used here 1e-10 and not again 1e-5?

Reply:
unclear ;)